File: api/app/services/custom_role_service.py
"""
Custom role service for organization role management.
"""
import json
from typing import List, Optional
from sqlalchemy.orm import Session

from app.services.base import BaseService
from app.repositories.custom_role_repository import CustomRoleRepository
from app.repositories.organization_repository import OrganizationMemberRepository
from app.models.custom_role import CustomRole
from app.models.organization import MemberRole
from app.schemas.custom_role import CustomRoleCreate, CustomRoleUpdate


class CustomRoleService(BaseService):
    """Service for custom role management."""

    # Default permissions for preset roles
    DEFAULT_ROLE_PERMISSIONS = {
        "管理员": [
            "apartment:view", "apartment:create", "apartment:edit",
            "room:view", "room:create", "room:edit",
            "tenant:view", "tenant:create", "tenant:edit",
            "lease:view", "lease:create", "lease:edit",
            "bill:view", "bill:create", "bill:edit",
            "utility:view", "utility:create", "utility:edit",
            "report:view", "report:export",
        ],
        "财务": [
            "apartment:view",
            "room:view",
            "tenant:view",
            "lease:view",
            "bill:view", "bill:create", "bill:edit",
            "utility:view",
            "report:view", "report:export",
        ],
        "运营": [
            "apartment:view", "apartment:create", "apartment:edit",
            "room:view", "room:create", "room:edit",
            "tenant:view", "tenant:create", "tenant:edit",
            "lease:view", "lease:create", "lease:edit",
            "bill:view",
            "utility:view", "utility:create", "utility:edit",
            "report:view", "report:export",
        ],
    }

    def __init__(self, db: Session):
        super().__init__(db)
        self.role_repo = CustomRoleRepository(db)
        self.member_repo = OrganizationMemberRepository(db)

    def list_roles(self, org_id: str, active_only: bool = True) -> List[CustomRole]:
        """List all custom roles in an organization."""
        return self.role_repo.find_by_organization(org_id, active_only)

    def get_role(self, role_id: str, org_id: str) -> Optional[CustomRole]:
        """Get a custom role by ID within an organization."""
        role = self.role_repo.get(role_id)
        if role and role.organization_id == org_id:
            return role
        return None

    def create_role(self, org_id: str, data: CustomRoleCreate) -> CustomRole:
        """
        Create a new custom role.

        Args:
            org_id: Organization ID
            data: Role creation data

        Returns:
            Created role

        Raises:
            ValueError: If role name already exists
        """
        # Check for duplicate name
        existing = self.role_repo.find_by_name(org_id, data.name)
        if existing:
            raise ValueError("角色名称已存在")

        # Convert permissions list to JSON string
        permissions_str = json.dumps(data.permissions) if data.permissions else None

        role = CustomRole(
            organization_id=org_id,
            name=data.name,
            description=data.description,
            is_system=False,
            is_active=True,
            permissions=permissions_str,
        )
        return self.role_repo.create(role)

    def update_role(
        self, role_id: str, org_id: str, data: CustomRoleUpdate
    ) -> Optional[CustomRole]:
        """
        Update a custom role.

        Args:
            role_id: Role ID
            org_id: Organization ID
            data: Update data

        Returns:
            Updated role or None if not found

        Raises:
            ValueError: If role is system preset or name conflict
        """
        role = self.get_role(role_id, org_id)
        if not role:
            return None

        if role.is_system:
            raise ValueError("系统预置角色不可修改")

        # Check for duplicate name if changing
        if data.name and data.name != role.name:
            existing = self.role_repo.find_by_name(org_id, data.name)
            if existing:
                raise ValueError("角色名称已存在")

        update_data = data.model_dump(exclude_unset=True)

        # Handle permissions conversion
        if "permissions" in update_data:
            update_data["permissions"] = json.dumps(update_data["permissions"])

        for field, value in update_data.items():
            if value is not None:
                setattr(role, field, value)

        self.db.commit()
        self.db.refresh(role)
        return role

    def delete_role(self, role_id: str, org_id: str) -> bool:
        """
        Delete a custom role.

        Args:
            role_id: Role ID
            org_id: Organization ID

        Returns:
            True if deleted

        Raises:
            ValueError: If role is system preset
        """
        role = self.get_role(role_id, org_id)
        if not role:
            return False

        if role.is_system:
            raise ValueError("系统预置角色不可删除")

        return self.role_repo.delete(role_id)

    def initialize_default_roles(self, org_id: str) -> List[CustomRole]:
        """
        Initialize default custom roles for a new organization.

        Args:
            org_id: Organization ID

        Returns:
            List of created roles
        """
        created_roles = []

        for name, permissions in self.DEFAULT_ROLE_PERMISSIONS.items():
            existing = self.role_repo.find_by_name(org_id, name)
            if existing:
                continue

            role = CustomRole(
                organization_id=org_id,
                name=name,
                description=f"系统预置{name}角色",
                is_system=True,
                is_active=True,
                permissions=json.dumps(permissions),
            )
            self.db.add(role)
            created_roles.append(role)

        if created_roles:
            self.db.commit()
            for role in created_roles:
                self.db.refresh(role)

        return created_roles

    def get_role_permissions(self, role_id: str, org_id: str) -> List[str]:
        """
        Get permissions for a role.

        Args:
            role_id: Role ID
            org_id: Organization ID

        Returns:
            List of permission codes
        """
        role = self.get_role(role_id, org_id)
        if not role or not role.permissions:
            return []

        try:
            return json.loads(role.permissions)
        except (json.JSONDecodeError, TypeError):
            return []

    def get_member_effective_permissions(
        self, org_id: str, user_id: str
    ) -> List[str]:
        """
        Get effective permissions for a member.

        Combines permissions from their role and custom role.

        Args:
            org_id: Organization ID
            user_id: User ID

        Returns:
            List of permission codes
        """
        membership = self.member_repo.find_membership(org_id, user_id)
        if not membership:
            return []

        # Owner and admin have all permissions
        if membership.role in [MemberRole.OWNER, MemberRole.ADMIN]:
            return ["*"]

        permissions = set()

        # Add permissions from custom role
        if membership.custom_role_id:
            custom_perms = self.get_role_permissions(
                membership.custom_role_id, org_id
            )
            permissions.update(custom_perms)

        return list(permissions)

File: api/src/services/customRole.service.ts
import type { CustomRole } from '@prisma/client';
import type { CustomRoleRepository } from '../repositories/customRole.repo.js';
import { defaultCustomRoleRepo } from '../repositories/customRole.repo.js';
import { createAppError } from '../utils/appError.js';
import { NotFoundMessages } from '../messages.js';
import { ulid } from 'ulid';
import { compareBooleanDesc, compareNaturalText } from '../utils/intuitiveSort.js';

/**
 * 创建角色输入
 */
export interface CreateCustomRoleInput {
  name: string;
  description?: string;
  permissions?: string;
}

/**
 * 更新角色输入
 */
export interface UpdateCustomRoleInput {
  name?: string;
  description?: string;
  permissions?: string;
  is_active?: boolean;
}

/**
 * 初始化默认角色
 */
const DEFAULT_ROLES = [
  { name: '管理员', description: '组织管理员' },
  { name: '财务', description: '财务人员' },
  { name: '运营', description: '运营人员' },
];

/**
 * CustomRole Service 接口
 */
export interface CustomRoleService {
  list(orgId: string): Promise<CustomRole[]>;
  initDefaultRoles(orgId: string): Promise<{ alreadyInitialized: boolean }>;
  create(orgId: string, data: CreateCustomRoleInput): Promise<CustomRole>;
  getById(orgId: string, id: string): Promise<CustomRole>;
  update(orgId: string, id: string, data: UpdateCustomRoleInput): Promise<CustomRole>;
  delete(orgId: string, id: string): Promise<void>;
}

/**
 * 创建 CustomRole Service 实例
 */
export function createCustomRoleService(
  getRepo: () => CustomRoleRepository = () => defaultCustomRoleRepo
): CustomRoleService {
  return {
    list: async (orgId: string) => {
      const roles = await getRepo().findByOrgId(orgId);
      return [...roles].sort(
        (left, right) =>
          compareBooleanDesc(left.is_active, right.is_active) ||
          compareNaturalText(left.name, right.name)
      );
    },

    initDefaultRoles: async (orgId: string) => {
      const existing = await getRepo().countByOrgId(orgId);
      if (existing > 0) {
        return { alreadyInitialized: true };
      }

      for (const d of DEFAULT_ROLES) {
        await getRepo().create({
          id: ulid().toLowerCase(),
          organization_id: orgId,
          name: d.name,
          description: d.description,
        });
      }

      return { alreadyInitialized: false };
    },

    create: async (orgId: string, data: CreateCustomRoleInput) => {
      return getRepo().create({
        id: ulid().toLowerCase(),
        organization_id: orgId,
        name: data.name,
        description: data.description,
        permissions: data.permissions,
      });
    },

    getById: async (orgId: string, id: string) => {
      const role = await getRepo().findByIdAndOrg(id, orgId);
      if (!role) {
        throw createAppError(404, NotFoundMessages.CUSTOM_ROLE);
      }
      return role;
    },

    update: async (orgId: string, id: string, data: UpdateCustomRoleInput) => {
      const existing = await getRepo().findByIdAndOrg(id, orgId);
      if (!existing) {
        throw createAppError(404, NotFoundMessages.CUSTOM_ROLE);
      }

      const updateData: Record<string, unknown> = {};
      if (data.name != null) updateData.name = data.name;
      if (data.description !== undefined) updateData.description = data.description;
      if (data.permissions !== undefined) updateData.permissions = data.permissions;
      if (data.is_active !== undefined) updateData.is_active = data.is_active;

      return getRepo().update(id, updateData);
    },

    delete: async (orgId: string, id: string) => {
      const existing = await getRepo().findByIdAndOrg(id, orgId);
      if (!existing) {
        throw createAppError(404, NotFoundMessages.CUSTOM_ROLE);
      }
      await getRepo().delete(id);
    },
  };
}

/**
 * 默认实例
 */
export const defaultCustomRoleService = createCustomRoleService();
